crates/cargo-util/src/process_builder.rs

```diff
@@ -340,9 +369,68 @@ impl ProcessBuilder {
         Ok(output)
     }
 
-    /// Converts `ProcessBuilder` into a `std::process::Command`, and handles the jobserver, if
-    /// present.
-    pub fn build_command(&self) -> Command {
+    /// Builds a command from `ProcessBuilder` and spawn it.
+    ///
+    /// There is a risk when spawning a process, it might hit "command line
+    /// too big" OS error. To handle those kind of OS errors, this method try
+    /// to reinvoke the command with a `@<path>` argfile that contains all the
+    /// arguments.
+    ///
+    /// * `apply`: Modify the command before invoking. Useful for updating [`Stdio`].
+    fn build_and_spawn(&self, apply: impl Fn(&mut Command)) -> io::Result<Child> {
+        let mut cmd = self.build_command();
+        apply(&mut cmd);
+
+        match cmd.spawn() {
+            Err(ref e) if self.should_retry_with_argfile(e) => {
+                let (mut cmd, _argfile) = self.build_command_with_argfile()?;
+                apply(&mut cmd);
+                cmd.spawn()
+            }
+            res => res,
+        }
+    }
+
+    /// Builds the command with an `@<path>` argfile that contains all the
+    /// arguments. This is primarily served for rustc/rustdoc command family.
+    ///
+    /// Ref:
+    ///
+    /// - https://doc.rust-lang.org/rustdoc/command-line-arguments.html#path-load-command-line-flags-from-a-path
+    /// - https://doc.rust-lang.org/rustc/command-line-arguments.html#path-load-command-line-flags-from-a-path>
+    fn build_command_with_argfile(&self) -> io::Result<(Command, NamedTempFile)> {
+        use std::io::Write as _;
+
+        let mut tmp = tempfile::Builder::new()
+            .prefix("cargo-argfile.")
+            .tempfile()?;
+
+        let path = tmp.path().display();
+        let mut cmd = self.build_command_without_args();
+        cmd.arg(format!("@{path}"));
```

Comment:
Observation: This won't break wrappers that don't understand argfiles because they already weren't spawning.